### ISSUE TITLE
Issues on Windows Server 2012 R2 but not Windows 7 or 10

### ISSUE THREAD

**Issue by mattduguid** [seanmcne/Microsoft.Xrm.Data.PowerShell#80]
Hi,

We've created an automation module for a service desk to perform CRM tasks and tested them as working 100% on Windows 7 or Windows 10 using the following,

- Microsoft Dynamics CRM SDK 2015
- Microsoft.Xrm.Data.Poewrshell v1.8

However when attempting to load the same on Windows Server 2012 R2 (a jumpbox for staff) we cannot connect CRM and it errors that certain DLL's (Microsoft.Xrm.Sdk.dll, etc) arent loaded but we can confirm they have been.

Has anyone else tried to use this on Windows 2012 R2?

Thanks

MD

**Comment by seanmcne** [comment]
Hey Matt - are you using windows powershell management framework 5 or whatever is included by default?   I tested using the default patched windows 2012 R2 box: 
1. Downloaded from releases here (since only v5 can use the install-module command) 
2. after saving the file locally, I right clicked it, I clicked properties, then unblocked the ZIP 
3. I placed the extracted folder under one of my powershell paths ($env:path), I placed it in the path: C:\windows\System32\WindowsPowershell\v1.0\Modules\
4. Run:  Import-Module Microsoft.Xrm.Data.PowerShell -verbose 

I can see the output now showing me all the dll's are loading - if I had to guess the zip was probably not unblocked at powershell is seeing the xrm sdk dll's as untrusted and therefore won't load them.

PS: my test was successful in loading the dll's and connecting to CRM

**Comment by seanmcne** [comment]
FYI here is my resulting module folder - C:\Windows\System32\WindowsPowerShell\v1.0\Modules\Microsoft.Xrm.Data.PowerShell\:
 
![image](https://cloud.githubusercontent.com/assets/2292260/13759894/63cf4670-e9fe-11e5-8f0d-672e80b2905f.png)

**Comment by mattduguid** [comment]
Hi Sean

Attached is a screenshot of the issue in action from a Windows Server 2012
R2. The server is using CRM SDK 2015 and CRM Powershell module 1.8 which is
same as the Windows 7 and 10 which work.

From the error message in screenshot and output from logging the assembly
binds (via Fusion) it appears the cause is a version mismatch on the dll.
Its trying to load v6.0.0.0 but the version in the modules folder and in
the CRM SDK are both v7.x.x.x?[image: windows2012r2_screen1.jpg]

HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Fusion
Add:
DWORD ForceLog set value to 1
DWORD LogFailures set value to 1
DWORD LogResourceBinds set value to 1
DWORD EnableLog set value to 1
String LogPath set value to folder for logs (e.g. C:\FusionLog\)

Thanks

MD

**Comment by seanmcne** [comment]
Thanks for the info Matt!  I went back and downloaded the 1.8 version (2.2 is the latest and what I originally tested with).  With this version I did the same test you did and it did function for me.  Can you make sure you do not have xrm tooling registered (we do not require it to be registered as a snapin, in fact it's better if you don't have that snapin registered as we essentially side-load it as a module). The module on github should be 100% contained at this point in time (you won't need to download or registere any other components).  It's possible that an older version of the xrm tooling snapin is registered and confusing powershell as its loading the dll's.  If you load the module using verbose, the first time you load it that will show you were it's loading the dll's from: 

![image](https://cloud.githubusercontent.com/assets/2292260/13779344/7922b1d0-ea88-11e5-889a-13fb0c48b5df.png)


Here are the file versions I have in my module directory: 

![image](https://cloud.githubusercontent.com/assets/2292260/13779215/ea37c186-ea87-11e5-9c2e-7dd935d15ea0.png)

**Comment by kenakamu** [comment]
I got same issue before and it was version conflict as you said, but I was loaded multiple versions in same session. Could you check if you remove all imported module first from user specific and common place, then import the latest once, please?

**Comment by mattduguid** [comment]
Thanks for the tips so far. By not loading the pssnapin and only the module the correct versions of the DLL's load from the module folder,

_#add-pssnapin microsoft.xrm.tooling.connector 
import-module Microsoft.Xrm.Data.PowerShell -verbose_

There is still a failure to connect because it cannot load the assembly Microsoft.IdentityModel.Clients.ActiveDirectory.dll but adding it manually from the SDK location works around it for now,

_Add-Type -Path "C:\crmsdk\SDK\Bin\Microsoft.IdentityModel.Clients.ActiveDirectory.dll"_

![windows2012r2_screen2](https://cloud.githubusercontent.com/assets/7642305/13799411/f1198ece-eb83-11e5-8871-f8ffb6c091a2.jpg)

After that we can successfully retrieve the list of the 5 CRM instances we have from the Windows 2012 R2 Server which previously failed.

Next we perform a Get-CrmConnection to the relevant CRM instance and here we see a new issue. The error returned is "more than 1 org was returned, retry with an exact org name", however when making the connection the parameter we provide for "organizationname" is only one of the unique names of one of the instances which works fine on Windows 7 and 10 as well,

![windows2012r2_screen3](https://cloud.githubusercontent.com/assets/7642305/13799456/5e2bb29e-eb84-11e5-908c-2196b66f8ce9.jpg)

**Comment by kenakamu** [comment]
Could you double check the binary version of the module? XrmTooling had this issue before and it should be fix in CRM2016 SDK. Version 2x of our module includes those.

**Comment by kenakamu** [comment]
close
